src/backend/optimizer/plan/transform.c

```diff
@@ -520,3 +522,87 @@ replace_sirvf_rte(Query *query, RangeTblEntry *rte)
 
 	return rte;
 }
+
+/*
+ * Does target list have SRFs?
+ */
+static
+bool tlist_has_srf(Query *query)
+{
+	if (query->hasTargetSRFs)
+	{
+		return true;
+	}
+
+	if (expression_returns_set( (Node *) query->targetList))
+	{
+		return true;
+	}
+
+	return false;
+}
+
+/*
+ * DISTINCT/DISTINCT ON/ORDER BY optimization.
+ * Remove DISTINCT clause if possibile, ex:
+ * select DISTINCT count(a) from t; to
+ * select count(a) from t;
+ * There is one row returned at most, DISTINCT and/or ON is pointless then.
+ * The same with ORDER BY clause;
+ */
+Query *remove_distinct_sort_clause(Query *parse)
+{
+	if (parse->hasAggs &&
+		parse->groupClause == NIL &&
+		!contain_mutable_functions((Node *) parse) &&
+		!tlist_has_srf(parse))
+	{
+		List	   *useless_tlist = NIL;
+		List	   *tles;
+		List	   *sortops;
+		List	   *eqops;
+		ListCell   *lc;
+
+		if (parse->distinctClause != NIL)
+		{
```

Comment:
Could your merge the `parse->distinctClause` and `parse->sortClause` into one list cluase , then call `get_sortgroupclauses_tles`? The two `foreach` blocks look same.

Reply:
It's intended, there may be either one or both, sometimes we could remove clause but sometimes not, and process fields like hasDistinctOn and etc.
See codes for detail.

Reply:
Yes, this looks good, not a big deal.